atomic_reactor/tasks/common.py

```diff
@@ -0,0 +1,92 @@
+"""
+Copyright (c) 2021 Red Hat, Inc
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+
+import abc
+import logging
+from dataclasses import dataclass, field, fields
+from typing import List, Dict, Any
+
+from atomic_reactor import util
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class PluginsDef:
+    """Defines the plugins to be executed by a task."""
+
+    prebuild: List[dict] = field(default_factory=list)
+    build: List[dict] = field(default_factory=list)
+    prepublish: List[dict] = field(default_factory=list)
+    postbuild: List[dict] = field(default_factory=list)
+    exit: List[dict] = field(default_factory=list)
+
+    def __post_init__(self):
+        """Validate the plugin definition right after the instance is created."""
+        to_validate = {
+            "prebuild_plugins": self.prebuild,
+            "buildstep_plugins": self.build,
+            "prepublish_plugins": self.prepublish,
+            "postbuild_plugins": self.postbuild,
+            "exit_plugins": self.exit,
+        }
+        util.validate_with_schema(to_validate, "schemas/plugins.json")
+
+
+@dataclass(frozen=True)
+class TaskParams:
+    """Task parameters (coming from CLI arguments)."""
+
+    build_dir: str
+    context_dir: str
+    config_file: str
+    user_params: Dict[str, Any]
+
+    # Note: do not give any attributes in this class default values, that would make dataclass
+    #   inheritance difficult. If they should have defaults, define them in the CLI parser.
+
+    @classmethod
+    def from_cli_args(cls, args: dict):
+        """Create a TaskParams instance from CLI arguments."""
+        args = cls._drop_known_unset_args(args)
+        params_str = args.pop("user_params", None)
+        params_file = args.pop("user_params_file", None)
+
+        if params_str:
+            user_params = util.read_yaml(params_str, "schemas/user_params.json")
+        elif params_file:
+            user_params = util.read_yaml_from_file_path(params_file, "schemas/user_params.json")
+        else:
+            user_params = {}
```

Comment:
Now that I think about it, should be this be allowed? The schema marks a bunch of properties as [required](https://github.com/containerbuildsystem/atomic-reactor/blob/master/atomic_reactor/schemas/user_params.json#L89). IMO none of those make sense anymore, but if we're going by the current schema then these `user_params` are invalid.

Reply:
let's disallow it, to find issues with incorrectly defined user_params caused by refactoring.

Will it make unittests harder?

Reply:
Hopefully not, if you need a `TaskParams` instance for unit tests you can create one directly, you don't need to go through `from_cli_args`. And the unit tests for the `common` module won't be harder either, I'm mocking the read_yaml methods anyway. Only unit tests for the parser (if I make one of the user params arguments required) will be slightly uglier.

Reply:
Or I can just raise an error here and leave the params as optional in the CLI, we may one day reach a point where --user-params can be empty

Reply:
Did that, keeping them optional in the CLI, lmk what you think

Reply:
+1

Reply:
@rcerven on the topic of user params being required. Do you think we should have them as required in the CLI for now? I'm thinking we might eventually refactor them away, so I only put a check in the TaskParams class and not in the CLI parser as that was easier to unit test.